lib/improver/nbhood.py

```diff
@@ -246,73 +305,171 @@ def _apply_kernel_for_smoothing(self, cube, ranges):
+
+
+class NeighbourhoodProcessing(object):
+    """
+    Apply a neigbourhood processing method to a thresholded cube.
+
+    When applied to a thresholded probabilistic cube, it acts like a
+    low-pass filter which reduces noisiness in the probabilities.
+
+    The neighbourhood methods will presently only work with projections in
+    which the x grid point spacing and y grid point spacing are constant
+    over the entire domain, such as the UK national grid projection
+
+    """
+
+    def __init__(self, neighbourhood_method, radii_in_km, lead_times=None,
+                 unweighted_mode=False):
+        """
+        Create a neighbourhood processing plugin that applies a smoothing
+        to points in a cube.
+
+        Parameters
+        ----------
+
+        neighbourhood_method : str
+            Name of the neighbourhood method to use. Options: 'circular'.
+        radii_in_km : float or List (if defining lead times)
+            The radii in kilometres of the neighbourhood to apply.
+            Rounded up to convert into integer number of grid
+            points east and north, based on the characteristic spacing
+            at the zero indices of the cube projection-x/y coords.
+        lead_times : None or List
+            List of lead times or forecast periods, at which thel radii
+            within radii_in_km are defined. The lead times are expected
+            in hours.
+        unweighted_mode : boolean
+            If True, use a circle with constant weighting.
+            If False, use a circle for neighbourhood kernel with
+            weighting decreasing with radius.
+
+        """
+        methods = {
+            "circular": CircularNeighbourhood}
+        try:
+            self.neighbourhood_method = methods[neighbourhood_method]
+        except KeyError:
+            msg = ("The neighbourhood_method requested: {} is not a "
+                   "supported method. Please choose from: {}".format(
+                       neighbourhood_method, methods.keys()))
+            raise KeyError(msg)
+
+        if isinstance(radii_in_km, list):
+            self.radii_in_km = [float(x) for x in radii_in_km]
+        else:
+            self.radii_in_km = float(radii_in_km)
+        self.lead_times = lead_times
+        if self.lead_times is not None:
+            if len(radii_in_km) != len(lead_times):
+                msg = ("There is a mismatch in the number of radii "
+                       "and the number of lead times. "
+                       "Unable to continue due to mismatch.")
+                raise ValueError(msg)
+        self.unweighted_mode = bool(unweighted_mode)
+
+    def __str__(self):
+        result = ('<NeighbourhoodProcessing: neighbourhood_method: {}; ' +
+                  'radii_in_km: {}; lead_times: {};' +
+                  'unweighted_mode: {}>')
+        return result.format(
+            self.neighbourhood_method, self.radii_in_km, self.lead_times,
+            self.unweighted_mode)
```

Comment:
Changed to be a `__repr__`. I've added a test for each of the `__repr__` methods.